### PR TITLE
Fix/303 use select2 in tags field

### PR DESCRIPTION
**Pull request by pixeljunkie**
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed issues  | chameleon-system/chameleon-system#303
| License       | MIT

The old tagsInput jquery plugin didn't work anymore and is now replaced with the select2 component.

### CODE REVIEW

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php [hunk 1]
```diff
@@ -406,7 +409,7 @@ public function HasContent()
         return $bHasContent;
     }
 
-    protected function GetCleanedTagString()
+    protected function GetCleanedTagString(): string
```
Comment (umulmrum) [inline src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: This is a BC break we should avoid.

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php [hunk 2]
```diff
@@ -91,21 +132,36 @@ public function DefineInterface()
...
+        $baseLanguageId = $cmsConfig->fieldTranslationBaseLanguageId;
+        $activeLanguageId = TdbCmsUser::GetActiveUser()->GetCurrentEditLanguageID();
+        if ($baseLanguageId !== $activeLanguageId) {
+            return '__'.$this->getLanguageService()->getLanguageIsoCode($activeLanguageId);
+        }
+
+        return '';
+    }
+
+    public function getAutocompleteTagList()
```
Comment (umulmrum) [inline src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]:
```suggestion
    public function getAutocompleteTagList(): array
```

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php [hunk 3]
```diff
@@ -118,97 +174,88 @@ public function GetTagList()
...
-                $aCurrentTags = explode(',', $sCurrentTags);
-                if (count($aCurrentTags) > 0) {
-                    $sTagBlackList = '';
-                    foreach ($aCurrentTags as $sCurrentTag) {
-                        $sTagBlackList .= "'".MySqlLegacySupport::getInstance()->real_escape_string(trim(strtolower($sCurrentTag)))."',";
-                    }
-                    $sTagBlackList = substr($sTagBlackList, 0, -1);
+        $inputFilter = $this->getInputFilterUtil();
+
+        $searchTerm = $inputFilter->getFilteredInput('q');
```
Comment (umulmrum) [inline src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: While we're at it, we could restrict this to GET parameters.

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php [hunk 4]
```diff
@@ -118,97 +174,88 @@ public function GetTagList()
...
+        if (null === $searchTerm) {
+            return [];
+        }
+
+        $searchTerm = strtolower(trim($searchTerm));
+
+        $sLanguageIsoName = $this->GetLanguageSuffix();
+        $sTagBlackListQueryPart = '';
+
+        $currentTags = $inputFilter->getFilteredInput('currentTags');
```
Comment (umulmrum) [inline src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: As above, please restrict to GET. In addition we could use the empty string as default value so the check in the following line gets easier.

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php [hunk 5]
```diff
@@ -19,7 +22,7 @@ class TCMSFieldLookupMultiselectTags extends TCMSFieldLookupMultiselect
     /**
      * if true, the field tries to load tag suggestions.
      *
-     * @var bool - default true
+     * @var bool - default false
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: removed

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php [hunk 6]
```diff
@@ -54,32 +57,70 @@ class TCMSFieldLookupMultiselectTags extends TCMSFieldLookupMultiselect
...
+            $html .= '<p id="'.TGlobal::OutHTML($this->name).'_suggestions" class="mt-2 tagSuggestions">
+<span class="label">'.TGlobal::OutHTML(TGlobal::Translate('chameleon_system_core.field_lookup_multi_select_tag.suggestions')).': </span>
+<span class="tagSuggestionList"></span>
+</p>';
         }
 
         return $html;
     }
 
+    private function getAjaxUrlParameter(): array
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: changed

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php (hunk 1 above)
Comment (pixeljunkie) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: It is? But it always return string. removed it and added docblock instead.

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php (hunk 2 above)
Comment (pixeljunkie) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: added

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php (hunk 3 above)
Comment (pixeljunkie) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: done

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php (hunk 4 above)
Comment (pixeljunkie) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: done

File: src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php (hunk 1 above)
Comment (umulmrum) [reply src/CoreBundle/private/library/classes/TCMSFields/TCMSFieldLookupMultiselectTags.class.php]: The BC breaks is that descendants of this class that override this method would be incompatible without also adding the return type declaration.

**Comment by umulmrum** [comment]
Thanks @pixeljunkie

**Pull request merged**